api/src/apps/apps.rs

```diff
@@ -48,19 +48,78 @@ use yansi::Paint;
 pub struct AppsService {
     config: Config,
     infrastructure: Box<dyn Infrastructure>,
-    apps_in_deployment: Mutex<HashSet<String>>,
+    app_guards: Mutex<HashMap<AppName, Arc<AppGuard>>>,
     web_host_meta_cache: Mutex<SizedCache<String, WebHostMeta>>,
 }
 
-struct DeploymentGuard<'a, 'b> {
-    apps_service: &'a AppsService,
-    app_name: &'b String,
+type GuardedResult = Result<Vec<Service>, AppsServiceError>;
+
+#[derive(Debug, Copy, Clone, PartialEq)]
+enum AppGuardKind {
+    Deployment,
+    Deletion,
+}
```

Comment:
Is it impossible for a deployment to race a deletion? What if the external request A wants to delete an app, and B wants to deploy that app (e.g. triggered by a subsequent build). Maybe it would be simpler to just cover all mutations to the same app using the same guard?

Reply:
That is true that this is an issue but a deployment or deletion should be atomic because they could interfere each other, leading to broken states of the application.

I think that this should be covered in a separate issue, introducing a background thread that deploys/deletes the applications. This thread should use a queue can handle the case of many subsequent deployments/deletions. 

I'll create one and reference it in the code.

Reply:
I created #40 for this.